src/adapters/morpho-aave-v2/common/morphoBasePoolAdapter.ts

```diff
@@ -0,0 +1,633 @@
+      )
+      const protocolTokenPromise = getTokenMetadata(
+        marketAddress,
+        this.chainId,
+        this.provider,
+      )
+      const [protocolToken, underlyingToken] = await Promise.all([
+        protocolTokenPromise,
+        underlyingTokenPromise,
+      ])
+
+      metadataObject[protocolToken.address] = {
+        protocolToken,
+        underlyingToken,
+      }
+    })
+
+    await Promise.all(promises)
+
+    return metadataObject
+  }
+
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    return Object.values(await this.buildMetadata()).map(
+      ({ protocolToken }) => protocolToken,
+    )
+  }
+
+  protected async fetchProtocolTokenMetadata(
+    protocolTokenAddress: string,
+  ): Promise<Erc20Metadata> {
+    const { protocolToken } = await this.fetchPoolMetadata(protocolTokenAddress)
+
+    return protocolToken
+  }
+
+  private async fetchPoolMetadata(protocolTokenAddress: string) {
+    const poolMetadata = (await this.buildMetadata())[protocolTokenAddress]
+
+    if (!poolMetadata) {
+      logger.error({ protocolTokenAddress }, 'Protocol token pool not found')
+      throw new Error('Protocol token pool not found')
+    }
+
+    return poolMetadata
+  }
+
+  protected async getUnderlyingTokenBalances({
+    protocolTokenBalance,
+  }: {
+    userAddress: string
+    protocolTokenBalance: TokenBalance
+    blockNumber?: number
+  }): Promise<Underlying[]> {
+    const { underlyingToken } = await this.fetchPoolMetadata(
+      protocolTokenBalance.address,
+    )
+
+    const underlyingTokenBalance = {
+      ...underlyingToken,
+      balanceRaw: protocolTokenBalance.balanceRaw,
+      type: TokenType.Underlying,
+    }
+
+    return [underlyingTokenBalance]
+  }
+
+  protected async fetchUnderlyingTokensMetadata(
+    protocolTokenAddress: string,
+  ): Promise<Erc20Metadata[]> {
+    const { underlyingToken } = await this.fetchPoolMetadata(
+      protocolTokenAddress,
+    )
+
+    return [underlyingToken]
+  }
+
+  async getPositions({
+    userAddress,
+    blockNumber,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const lensContract = MorphoAaveV2Lens__factory.connect(
+      this.lensAddress,
+      this.provider,
+    )
+    const tokens = await this.getProtocolTokens()
+    const positionType = this.getProtocolDetails().positionType
+
+    const getBalance = async (
+      market: Erc20Metadata,
+      userAddress: string,
+      blockNumber: number,
+    ): Promise<number> => {
+      let balanceRaw
+      if (positionType === PositionType.Supply) {
+        ;[, , balanceRaw] = await lensContract.getCurrentSupplyBalanceInOf(
+          market.address,
+          userAddress,
+          { blockTag: blockNumber },
+        )
+      } else {
+        ;[, , balanceRaw] = await lensContract.getCurrentBorrowBalanceInOf(
+          market.address,
+          userAddress,
+          { blockTag: blockNumber },
+        )
+      }
+      return Number(balanceRaw)
```

Comment:
is the Number type required ? Its not safe at all, you should consider working with `bigint` as must as possible